### PR TITLE
File CDK: remove default file based cursor

### PR DESCRIPTION
**Pull request by erohmensing**
## What
* Remove the `DefaultFileBasedCursor` in favor of all file-based streams getting a `FileBasedConcurrentCursor`
  * remove `AbstractFileBasedCursor`
* All file based streams now emit final state messages on full refresh (helpful for record counts)
* All file based streams now support concurrency (I think)?
* Move incrementally slowly toward merging all file based/concurrent combinations into one

## How
Steps: 
1. Remove `DefaultFileBasedCursor`. Update `InMemoryFileSource` to default to a `FileBasedConcurrentCursor` instead.
2. ⚠️ remove the original `incremental_scenarios.py`, as it contained replicas of a subset of `concurrent_incremental_scenarios.py` with the default cursor instead of the defined concurrent one. 
3. ⚠️ rename `concurrent_incremental_scenarios.py` to `incremental_scenarios.py`, unset the specific cursor setting, and rename the tests accordingly. Marking this with a warning as it is a bit confusing in the diff, which makes it look like we just removed `concurrent_incremental_scenarios.py`, when that's not the case.
4. Update the other tests' expected records, as the default cursor, `FileBasedConcurrentCursor` emits a final state message, like we are expecting of all sources by now.
5. Remove `LowHistoryLimitCursor`; all low history limit cursors are `LowHistoryLimitConcurrentCursor`s accordingly. 
6. Update branching logic to implement assertions as all file-based sources should have concurrency and abstractconcurrentfilebasedcursors
7. Make sure all type references accurately reflect that an AbstractConcurrentFileBasedCursor is expected. Thus, delete AbstractFileBasedCursor after merging it with AbstractConcurrentFileBasedCursor

## Review guide
Follow the implementation steps for best results

## User Impact

We're removing some things that sources might use. If we find sources that do this, we will pin their CDK versions and create issues to move them to the new interfaces.

TODO:
Update or pin the max CDK version of the following connectors: 
- [ ] Connectors using DefaultFileBasedCursor
  * `source-azure-blob-storage` (alpha/community)
  * `source-gcs` (alpha/community)
  * `source-google-drive` (alpha/community)
  * `source-microsoft-onedrive` (alpha/community)
  * `souce-microsoft-sharepoint` (alpha/certified)
  * `source-s3` (ga/certified)
- [x] Look for usage of LowHistoryLimitCursor - None
- [x] Look for usage of AbstractFIleBasedCursor - None

## Can this PR be safely reverted and rolled back?

- [x] YES 💚
- [ ] NO ❌

### CODE REVIEW

File: airbyte-cdk/python/unit_tests/sources/file_based/scenarios/incremental_scenarios.py
Comment (erohmensing) [inline airbyte-cdk/python/unit_tests/sources/file_based/scenarios/incremental_scenarios.py:1]: The diffs in this file come from it being
* delete incremental_scenarios.py
* clean up concurrent_incremental_scenarios.py and rename it

I originally did this in 2 separate prs for the diffs, but all the changes should go in together. This still might be confusing though - i can resplit it for review and then re-merge if that helps.

File: airbyte-cdk/python/unit_tests/sources/file_based/scenarios/incremental_scenarios.py
Comment (clnoll) [inline airbyte-cdk/python/unit_tests/sources/file_based/scenarios/incremental_scenarios.py:1353]: Did this need to be reordered to make the test pass? IIRC there was at least one test where they were intentionally set out-of-order. I don't know if it was this specific test, but there was one that helped identify what would have been a bug in the order in which items get popped off of history.

File: airbyte-cdk/python/unit_tests/sources/file_based/scenarios/incremental_scenarios.py
Comment (clnoll) [reply airbyte-cdk/python/unit_tests/sources/file_based/scenarios/incremental_scenarios.py:1]: I'm fine with them together. Did a pass through and it looks like everything deleted from the concurrent incremental is accounted for here.

**Review by clnoll — APPROVED** [review]
Nice @erohmensing! Just one question.

File: airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/concurrent/cursor/abstract_concurrent_file_based_cursor.py
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/concurrent/cursor/abstract_concurrent_file_based_cursor.py:21]: nit: not something that needs to block this PR, but we should figure out what naming scheme we want in the future. Should this just be an `AbstractFileBasedCursor`?

File: airbyte-cdk/python/airbyte_cdk/sources/file_based/file_based_source.py
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/file_based/file_based_source.py:188]: flagging the todo. Can we either use the stream facade or remove the todo if it doesn't matter?

I suspect we can't just use the facade because of the weird set up for the check

**Review by girarda — APPROVED** [review]
approved pending answering Catherine's question on the test setup https://github.com/airbytehq/airbyte/pull/37488/files#r1576905289

File: airbyte-cdk/python/unit_tests/sources/file_based/scenarios/incremental_scenarios.py
Comment (erohmensing) [reply airbyte-cdk/python/unit_tests/sources/file_based/scenarios/incremental_scenarios.py:1353]: I didn't reorder anything, so this might have been a weird diff in the old incremental_scenarios to the new incremental_scenarios. I will take a further look to check though!

**Pull request closed without merging**
